### PR TITLE
Destination Snowflake: lowered eager flush threshold and added runtime calc

### PR DESCRIPTION
**Pull request by davinchia**
## What
Eagerly flushes when the used memory or tracked memory hits 70%. The used memory happens more in low memory containers (e.g. 1 Gi) and because the record estimation is not always accurate. This prevents scenarios where OOM. The next step is to account for the jump in memory when processing a Batch. This is roughly 1.5x the batch size but should be more accurately measured

The longer-term goal is to use a memory profiler like VisualVM

## How
Adds another trigger for when the AsyncConsumer determines a queue is ready to flush. This is due to a delta between Global Memory Manager's measured allocated memory and what the `Runtime` informs us

## Recommended reading order
1. `DetectStreamToFlush.java`

## 🚨 User Impact 🚨
No breaking changes, this will address the memory issues we were initially seeing with Snowflake

*For connector PRs, use this section to explain which type of semantic versioning bump occurs as a result of the changes. Refer to our [Semantic Versioning for Connectors](https://docs.airbyte.com/contributing-to-airbyte/#semantic-versioning-for-connectors) guidelines for more information. **Breaking changes to connectors must be documented by an Airbyte engineer (PR author, or reviewer for community PRs) by using the [Breaking Change Release Playbook](https://docs.google.com/document/d/1VYQggHbL_PN0dDDu7rCyzBLGRtX-R3cpwXaY8QxEgzw/edit).***

*If there are breaking changes, please merge this PR with the 🚨🚨 emoji so changelog authors can further highlight this if needed.*

## Pre-merge Actions
*Expand the relevant checklist and delete the others.*

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- Connector version is set to `0.0.1`
    - `Dockerfile` has version `0.0.1`
- Documentation updated
    - Connector's `README.md`
    - Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - `docs/integrations/<source or destination>/<name>.md` including changelog with an entry for the initial version. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - `docs/integrations/README.md`

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- Unit & integration tests added

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- Create a non-forked branch based on this PR and test the below items on it
- Build is successful
- If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).

</details>

<details><summary><strong>Connector Generator</strong></summary>

- Issue acceptance criteria met
- PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests#pull-request-title-convention)
- If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- Documentation which references the generator is updated as needed

</details>

### CODE REVIEW

**Comment by ryankfu** [comment]
@davinchia thanks for creating this branch. I will need to push up the changes to the `computeQueueThreshold` I had locally and remove the portion in `SnowflakeDestination` that uses the AsyncConsumer

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/DetectStreamToFlush.java [hunk 1]
```diff
@@ -69,10 +69,15 @@ public Optional<StreamDescriptor> getNextStreamToFlush() {
...
+    // Eagerly flush when the Runtime used memory is 70% of the total memory. This is to avoid cases
+    // where
+    // the record estimation is inaccurate and we end up using too much memory.
+    // TODO: (ryankfu) - We should improve the record estimation to avoid this.
+    final boolean isBuffer70Full =
+        EAGER_FLUSH_THRESHOLD <= Math.max((double) bufferDequeue.getTotalGlobalQueueSizeBytes() / bufferDequeue.getMaxQueueSizeBytes(),
+            (double) (Runtime.getRuntime().totalMemory() - Runtime.getRuntime().freeMemory()) / Runtime.getRuntime().totalMemory());
     // when we are closing or queues are very full, flush regardless of how few items are in the queue.
-    return isClosing.get() || isBuffer90Full ? 0 : QUEUE_FLUSH_THRESHOLD_BYTES;
+    return isClosing.get() || isBuffer70Full ? 0 : QUEUE_FLUSH_THRESHOLD_BYTES;
```
Comment (edgao) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/DetectStreamToFlush.java:80]: nit: rename to `isBufferTooFull`

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java [hunk 2]
```diff
@@ -150,7 +150,7 @@ public List<PartialAirbyteMessage> flushStates() {
       while (true) {
         final Long oldestState = stateIdQueue.peek();
         final boolean emptyQ = oldestState == null;
-        final boolean noCorrespondingStateMsg = stateIdToState.get(oldestState) == null;
+        final boolean noCorrespondingStateMsg = !emptyQ && stateIdToState.get(oldestState) == null;
```
Comment (ryankfu) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/state/GlobalAsyncStateManager.java:153]: This fixes a NPE, not related to memory constraints

File: airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/RunningFlushWorkers.java [hunk 3]
```diff
@@ -12,16 +12,24 @@
...
  */
+@Slf4j
 public class RunningFlushWorkers {
 
   private final ConcurrentMap<StreamDescriptor, ConcurrentMap<UUID, Optional<Long>>> streamToFlushWorkerToBatchSize;
+  private final ScheduledExecutorService debugLoop;
 
   public RunningFlushWorkers() {
     streamToFlushWorkerToBatchSize = new ConcurrentHashMap<>();
+    debugLoop = Executors.newSingleThreadScheduledExecutor();
```
Comment (edgao) [inline airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/destination_async/RunningFlushWorkers.java:31]: let's make this a daemon thread
```suggestion
    debugLoop = ExecutorService exec = Executors.newFixedThreadPool(1,
        new ThreadFactory() {
            public Thread newThread(Runnable r) {
                Thread t = Executors.defaultThreadFactory().newThread(r);
                t.setDaemon(true);
                return t;
            }
        });
```

then you don't need to explicitly shut down the pool

**Review by edgao — APPROVED** [review]
couple nits but I wouldn't block this PR on them. :shipit:

**Review by cgardens — APPROVED** [review]

**Pull request closed without merging**